File: 124-Binary-Tree-Maximum-Path-Sum/0124-Binary-Tree-Maximum-Path-Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def find(self,root):
        if not root:
            return 0
        left = max(0, self.find(root.left))
        right = max(0, self.find(root.right))
        self.result = max(self.result, root.val + left + right)
        return max(left,right) + root.val
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.result = float('-inf')
        self.find(root)
        return self.result

File: 124-Binary-Tree-Maximum-Path-Sum/0124-Binary-Tree-Maximum-Path-Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int result = INT_MIN;
    int find(TreeNode* root){
        // base condition to return leaf node
        if(root == NULL){
            return 0;
        }
        // if sum is -ve we return 0
        int left = max(0, find(root->left));
        int right = max(0, find(root->right));
        // updating the maximum sum till now
        result = max(result, root->val + left + right);
        //returning the maximum possible sum
        return root->val + max(left, right);
    }
    int maxPathSum(TreeNode* root) {
        find(root);
        return result;   
    }
};
